pkg/cable/libreswan/libreswan.go

```diff
@@ -1,57 +1,377 @@
+
+				klog.Infof("Creating connection to %v", endpoint)
+				klog.Infof("Whacking with %v", args)
+
+				cmd = exec.Command("/usr/libexec/ipsec/whack", args...)
+				cmd.Stdout = os.Stdout
+				cmd.Stderr = os.Stderr
+
+				if err := cmd.Run(); err != nil {
+					switch err := err.(type) {
+					case *exec.ExitError:
+						klog.Errorf("Error adding a connection with args %v; got exit code %d: %v", args, err.ExitCode(), err)
+					default:
+						return "", fmt.Errorf("Error adding a connection with args %v: %v", args, err)
+					}
+				}
+
+				/*
+					if err := Route(endpoint.Spec.CableName); err != nil {
+						return "", fmt.Errorf("Error routing connection %v: %v", endpoint.Spec.CableName, err)
+					}
+
+					if err := Initiate(endpoint.Spec.CableName); err != nil {
+						return "", fmt.Errorf("Error initiating connection %v: %v", endpoint.Spec.CableName, err)
+					}
+				*/
+
+				cmd = exec.Command("/usr/libexec/ipsec/whack", "--route", "--name", connectionName)
+				cmd.Stdout = os.Stdout
+				cmd.Stderr = os.Stderr
+
+				if err := cmd.Run(); err != nil {
+					//return "", fmt
+					klog.Errorf("Error routing connection %s: %v", connectionName, err)
+				}
+
+				cmd = exec.Command("/usr/libexec/ipsec/whack", "--initiate", "--name", connectionName)
+				cmd.Stdout = os.Stdout
+				cmd.Stderr = os.Stderr
+
+				if err := cmd.Run(); err != nil {
+					//return "", fmt
+					klog.Errorf("Error initiating a connection with args %v: %v", args, err)
+				}
+			}
+		}
+	}
+
+	i.connections = append(i.connections, subv1.Connection{Endpoint: endpoint.Spec, Status: subv1.Connected})
+
+	return remoteEndpointIP, nil
 }
 
 // DisconnectFromEndpoint disconnects from the connection to the given endpoint.
-func (i *libreSwan) DisconnectFromEndpoint(endpoint types.SubmarinerEndpoint) error {
-	return fmt.Errorf("Not implemented")
+func (i *libreswan) DisconnectFromEndpoint(endpoint types.SubmarinerEndpoint) error {
+	leftSubnets := extractSubnets(i.localEndpoint.Spec)
+	rightSubnets := extractSubnets(endpoint.Spec)
+
+	klog.Infof("Deleting connection to %v", endpoint)
+
+	if len(leftSubnets) > 0 && len(rightSubnets) > 0 {
+		for lsi := range leftSubnets {
+			for rsi := range rightSubnets {
+				connectionName := fmt.Sprintf("%s-%d-%d", endpoint.Spec.CableName, lsi, rsi)
+
+				args := []string{}
+
+				args = append(args, "--delete")
+				args = append(args, "--name", connectionName)
+
+				klog.Infof("Whacking with %v", args)
+
+				cmd := exec.Command("/usr/libexec/ipsec/whack", args...)
+				cmd.Stdout = os.Stdout
+				cmd.Stderr = os.Stderr
+
+				if err := cmd.Run(); err != nil {
+					switch err := err.(type) {
+					case *exec.ExitError:
+						klog.Errorf("Error deleting a connection with args %v; got exit code %d: %v", args, err.ExitCode(), err)
+					default:
+						return fmt.Errorf("Error deleting a connection with args %v: %v", args, err)
+					}
+				}
+			}
+		}
+	}
+
+	for j := range i.connections {
```

Comment:
can we refactor it out to a separate method? (removeConnectionForEndpoint?)

Reply:
Done.